File: frontend/src/types/pipeline.ts
import type { GenerationResult } from "./gen";

/** Names of the three canonical pipeline stages, in execution order. */
export type PipelineStageName = "stats" | "features" | "gen";

/** Status of one stage entry in the per-stage report. */
export type PipelineStageStatus = "skipped" | "completed" | "failed";

/** One ordered stage entry of the POST /pipeline/numbers report (R2). */
export interface PipelineStageResult {
  name: PipelineStageName;
  status: PipelineStageStatus;
  snapshot_id: number | null;
  fingerprint: string | null;
  error_code: string | null;
  detail: string;
}

/**
 * Response data of POST /pipeline/numbers: the ordered three-stage report plus
 * the generation echo, which is null when any stage failed (S2 contract).
 */
export interface PipelineRunResult {
  stages: PipelineStageResult[];
  result: GenerationResult | null;
}

/** Payload for POST /pipeline/numbers. */
export interface PipelineRunParams {
  lottery_id: number;
  count?: number;
  seed?: number;
}

File: backend/src/backend/app/schemas/pipeline.py
"""Pydantic v2 schemas for the pipeline orchestrator surface (R3, D10).

One endpoint: ``POST /pipeline/numbers``. The request forbids unknown fields;
the response carries the ordered eight-stage report plus the generator output
(null on failure). Stage statuses are restricted to the R3 set.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from backend.app.schemas.gen import GenerationResult


class PipelineRunRequest(BaseModel):
    """Payload for ``POST /pipeline/numbers`` (unknown fields rejected).

    ``lottery_id`` is required; ``count`` defaults backend-side to 10
    (GEN-002) and ``seed`` stays optional (GEN-003).
    """

    model_config = ConfigDict(extra="forbid")

    lottery_id: int = Field(gt=0)
    count: int | None = None
    seed: int | None = None


class PipelineStageResult(BaseModel):
    """One ordered stage entry of the per-stage report (R3)."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    status: Literal["skipped", "completed", "failed"]
    snapshot_id: int | None = None
    fingerprint: str | None = None
    error_code: str | None = None
    detail: str = ""


class PipelineRunResult(BaseModel):
    """``POST /pipeline/numbers`` response data (R1/R3).

    ``stages`` lists exactly the eight canonical stages in order; ``result``
    is the generation echo or ``None`` when the run aborted.
    """

    model_config = ConfigDict(from_attributes=True)

    stages: list[PipelineStageResult]
    result: GenerationResult | None = None
